### PR TITLE
fix(race): close remaining #3247 shared-state races

### LINKED ISSUE #3247: [BUG] Systematic data races: 20+ extension global maps unprotected, wrong lock in SetProviderService, RouterChain.Route lockless
### Verification Checklist

- [x] I have searched the [existing Issues](https://github.com/apache/dubbo-go/issues) and believe this is not a duplicate

### Go Version

1.24

### Dubbo-go Version

v3.3.1

### OS

Linux / All platforms

### Bug Description

A comprehensive audit of the dubbo-go codebase reveals **44 data race conditions** across the framework. The most critical is a **systematic pattern** in the `common/extension/` package where 20+ global maps are accessed without any synchronization, plus a confirmed bug where `SetProviderService` uses the **wrong lock variable**. These issues will cause `fatal error: concurrent map read and map write` crashes in production.

Running `go test -race` on any test that exercises dynamic service registration/deregistration, configuration updates, or concurrent RPC calls will reproduce these races.

---

## Category 1: `common/extension/` — 20+ global maps with zero lock protection (P0)

Every file in `common/extension/` follows the exact same unsafe pattern:

```go
var things = make(map[string]func() Thing)

func SetThing(name string, v func() Thing) {
    things[name] = v  // write: no lock
}

func GetThing(name string) Thing {
    if things[name] == nil {  // read: no lock
        panic("...")
    }
    return things[name]()
}
```

**Affected files** (all in `common/extension/`):

| File | Global map(s) | Extra risk |
|------|--------------|------------|
| `filter.go` | `filters`, `rejectedExecutionHandler` | Has `UnregisterFilter` + `GetAllFilterNames` (runtime mutation) |
| `protocol.go` | `protocols` | Hot path: every invoker creation |
| `registry.go` | `registries` | |
| `cluster.go` | `clusters` | Hot path: invoker reorganization |
| `loadbalance.go` | `loadbalances` | Has `UnregisterLoadbalance` (runtime mutation) |
| `configurator.go` | `configurator` | Config center hot reload path |
| `config_center.go` | `configCenters` | |
| `config_center_factory.go` | `configCenterFactories` | |
| `config.go` | `configs` | |
| `metadata_report_factory.go` | `metaDataReportFactories` | |
| `proxy_factory.go` | `proxyFactories` | |
| `tps_limit.go` | `tpsLimitStrategy`, `tpsLimiter` | |
| `auth.go` | `authenticators`, `accessKeyStorages` | |
| `rest_client.go` | `restClients` | |
| `rest_server.go` | `restServers` | |
| `config_reader.go` | `configReaders`, `defaults` | `GetDefaultConfigReader` returns internal map reference! |
| `config_post_processor.go` | `processors` | |
| `router_factory.go` | `routers` | `GetRouterFactories` returns internal map reference! |
| `registry_directory.go` | `directories`, `defaultDirectory` | |
| `service_discovery.go` | `discoveryCreatorMap` | |
| `service_instance_selector_factory.go` | `serviceInstanceSelectorMappings` | |
| `logger.go` | `logs` | |
| `otel_trace.go` | `traceExporterMap` | Iterated in shutdown callback |

**Worst cases**: `GetDefaultConfigReader()` and `GetRouterFactories()` **return the internal map reference directly** — callers iterate the returned map without any lock, while `Set*` functions write to the same map concurrently. This is guaranteed to crash.

**Recommended fix**: Introduce a generic `Registry[T]` container with built-in `sync.RWMutex`:

```go
type Registry[T any] struct {
    mu    sync.RWMutex
    items map[string]T
}

func (r *Registry[T]) Set(name string, v T) {
    r.mu.Lock()
    defer r.mu.Unlock()
    r.items[name] = v
}

func (r *Registry[T]) Get(name string) (T, bool) {
    r.mu.RLock()
    defer r.mu.RUnlock()
    v, ok := r.items[name]
    return v, ok
}

func (r *Registry[T]) Snapshot() map[string]T {
    r.mu.RLock()
    defer r.mu.RUnlock()
    m := make(map[string]T, len(r.items))
    for k, v := range r.items {
        m[k] = v
    }
    return m
}
```

This would fix all 20+ files with a single refactor.

---

## Category 2: `dubbo.go` SetProviderService uses wrong lock (P0)

**File**: `dubbo.go:300-306`

```go
var (
    consumerServices = map[string]*client.ClientDefinition{}
    conLock          sync.RWMutex     // for consumerServices
    providerServices = map[string]*server.ServiceDefinition{}
    proLock          sync.RWMutex     // for providerServices
)

func SetProviderService(svc common.RPCService) {
    conLock.Lock()           // BUG: should be proLock!
    defer conLock.Unlock()
    providerServices[...] = ...  // writing providerServices with wrong lock
}
```

`SetProviderService` uses `conLock` (consumer lock) to protect writes to `providerServices`, but `loadProvider()` (L219-220) uses `proLock.RLock()` to protect reads of the same map. **Two different locks protecting the same map = no protection at all.** Concurrent `SetProviderService` and `loadProvider` will crash.

Also: `GetConsumerConnection` (L308-310) reads `consumerServices` without holding `conLock` at all.

**Fix**: Change L301 `conLock.Lock()` → `proLock.Lock()`. Add `conLock.RLock()` to `GetConsumerConnection`.

---

## Category 3: `RouterChain.Route` reads invokers without lock (P0)

**File**: `cluster/router/chain/chain.go:55-67, 91-98`

```go
func (c *RouterChain) Route(...) []base.Invoker {
    finalInvokers := make([]base.Invoker, 0, len(c.invokers))  // read: NO LOCK
    for _, invoker := range c.invokers {                        // iterate: NO LOCK
        // ...
    }
}

func (c *RouterChain) SetInvokers(invokers []base.Invoker) {
    c.mutex.Lock()           // write: has lock
    defer c.mutex.Unlock()
    c.invokers = invokers
}
```

`Route()` is the **hottest path** in the framework (called on every RPC), but reads `c.invokers` without any lock. `SetInvokers` writes with `mutex.Lock()`. Slice header tear (reading partially-updated data/len/cap) can cause out-of-bounds panic or routing to wrong nodes.

**Fix**: Add `c.mutex.RLock()` in `Route()`, snapshot the slice before iteration.

---

## Category 4: `config/service.go` returns internal map references (P0)

**File**: `config/service.go:91-93, 102-104`

```go
func GetProviderServiceMap() map[string]common.RPCService {
    return proServices  // returns internal map without lock or copy
}
```

Caller `provider_config.go:171` iterates the returned map with `for range`, while `SetProviderService` (L54-62, holding `proServicesLock`) writes to the same map concurrently. No lock held during iteration = crash.

**Fix**: Return a copy under lock.

---

## Category 5: `RegistryDirectory` fields accessed without locks (P1)

Multiple fields in `RegistryDirectory` are accessed concurrently without proper synchronization:

1. **`IsAvailable()` reads `cacheInvokers` without `invokersLock`** (L534-546) — `setNewInvokers()` writes with lock, but `IsAvailable()` reads without lock
2. **`Destroy()` reads/writes `cacheInvokers` without `invokersLock`** (L549-573)
3. **`configurators` slice appended in `convertUrl()` (L379) without lock**, read in `overrideUrl()` (L575-579) without lock
4. **`SubscribedUrl` written in goroutine (L165) without lock**, read in `Destroy()` (L559)

---

## Category 6: Other notable races (P1-P2)

| Issue | File | Description |
|-------|------|-------------|
| `rootConfig` global | `config/config_loader.go:37,56` | Written by `Load()`/`SetRootConfig()`, read by `GetRootConfig()` everywhere, no atomic/lock |
| `overrideSubscribeListener.configurator` | `registry/protocol/protocol.go:338-394` | Written in `Notify()` goroutine, read in `doOverrideIfNecessary()`, no lock |
| `hystrix Filter.res` map | `filter/hystrix/filter.go:130-178` | Written under `configLoadMutex.Lock()`, read in `hystrix.Do` callback **without lock** |
| `applicationName` global | `metrics/common.go:46-63` | Written by `InitAppInfo()`, read by metrics goroutine, no atomic |
| `versionInt` map × 2 | `protocol/dubbo/impl/hessian.go:166`, `hessian2/hessian_response.go:405` | Concurrent reads from multiple connection goroutines |
| `customizers` slice | `extension/service_instance_customizer.go:28-41` | `AddCustomizers` appends+sorts, `GetCustomizers` returns internal slice reference |
| `scheduledLookUp` goroutine | `registry/nacos/registry.go:233-238` | Uses `time.Sleep` without context, can't exit promptly on shutdown |
| `Subscribe` timeout goroutine leak | `registry/directory/directory.go:162-211` | Timeout returns error but goroutines continue running indefinitely |

---

### Summary

| Category | Count | Severity |
|----------|-------|----------|
| extension global maps (systematic) | 20+ maps across 22 files | P0 |
| Wrong lock (`SetProviderService`) | 1 | P0 |
| `RouterChain.Route` lockless | 1 | P0 |
| `config/service.go` internal map exposure | 2 maps | P0 |
| `RegistryDirectory` field races | 4 fields | P1 |
| Other races | 10+ | P1-P2 |

### Reproduction

```bash
go test -race ./...
```

Any test exercising concurrent service registration + RPC invocation will trigger race detector reports.

### Suggested Approach

1. **Introduce `extension.Registry[T]` generic container** — fixes 20+ files in one refactor
2. **Fix `dubbo.go:301` wrong lock** — one-line fix, highest ROI
3. **Add `RLock` to `RouterChain.Route`** — critical hot path
4. **Return map copies from `config/service.go`** — straightforward
5. **Add locks to `RegistryDirectory` field accesses** — systematic review needed

### LINKED ISSUE #3248: [Enhancement] Codebase quality improvements: URL redesign, extension safety, lifecycle management, API hygiene
## Background

After auditing the dubbo-go codebase for memory leaks (#3246) and data races (#3247), the root causes point to several structural issues that have accumulated over the project's 10-year history. This issue proposes a prioritized improvement roadmap.

---

## 1. Redesign `common.URL` — the core data structure needs a major overhaul

`common.URL` is the single most important type in the framework. Every invoker, exporter, registry, and configuration path revolves around it. But it has grown organically into a god object with serious problems:

**Current issues:**
- `params` (url.Values) exposed via `GetParams()` without lock or copy — callers mutate internals
- `attributes` (map[string]any) is write-only — no `DeleteAttribute`, no `ClearAttributes`
- `AddParam` vs `SetParam` semantic confusion (`Add` appends, not overwrites) causing silent accumulation
- `MergeURL` reads `anotherUrl.attributes` without `attributesLock` — data race
- `GetCacheInvokerMapKey` allocates a full temporary URL on every call — unnecessary GC pressure
- `SubURL` pointer creates implicit reference chains that prevent GC
- Multiple key generation methods (`Key()`, `GetCacheInvokerMapKey()`, `ServiceKey()`) with different formats, causing key mismatch bugs (e.g., invoker blacklist)

**Proposed redesign:**
1. Make `params` immutable after construction — provide `WithParam(k, v)` that returns a new URL (or at minimum, return copies from getters)
2. Add `DeleteAttribute(key)` and `ClearAttributes()` methods
3. Deprecate `AddParam` — it's almost always misused when `SetParam` is intended
4. Unify key generation into a single canonical `Key()` method
5. Consider splitting URL into `URLSpec` (immutable identity) + `URLState` (mutable runtime state)
6. Add `Clone()` that does proper deep copy of attributes

---

## 2. Make `common/extension/` concurrency-safe — introduce `Registry[T]`

The extension package has 20+ global maps, all following the same unsafe pattern (no locks, some returning internal map references). This is the single largest source of potential crashes.

**Proposed solution:** A generic thread-safe registry:

```go
package extension

type Registry[T any] struct {
    mu    sync.RWMutex
    items map[string]T
}

func NewRegistry[T any]() *Registry[T] { ... }
func (r *Registry[T]) Register(name string, v T) { ... }
func (r *Registry[T]) Get(name string) (T, bool) { ... }
func (r *Registry[T]) MustGet(name string) T { ... }  // panics if missing
func (r *Registry[T]) Unregister(name string) { ... }
func (r *Registry[T]) Snapshot() map[string]T { ... }  // returns copy
func (r *Registry[T]) Names() []string { ... }         // returns copy
```

This single abstraction would fix all 22 files in `common/extension/` and eliminate an entire class of bugs. It also standardizes the API (current code mixes `Set*`/`Get*`/`Unregister*`/`GetAll*` inconsistently).

---

## 3. Implement proper lifecycle management for invokers/exporters

Current state:
- `registryProtocol.Destroy()` doesn't clean `overrideListeners` or `serviceConfigurationListeners`
- `RegistryDirectory.Destroy()` doesn't clean `cacheInvokersMap`
- Invoker blacklist entries become permanently orphaned due to key mismatch
- `exporterChangeableWrapper` URLs are set *after* being stored in the map, creating a race window
- No `Closeable`/`Lifecycle` interface enforced — each component invents its own cleanup pattern

**Proposed improvements:**
1. Define a `Lifecycle` interface: `Start()`, `Stop()`, `IsRunning() bool`
2. Enforce cleanup contracts: every `Store` must have a corresponding `Delete` path
3. Add integration tests that verify zero goroutine/memory leaks after full startup-shutdown cycle
4. Use `context.Context` consistently for cancellation (replace `time.Sleep` polling patterns in nacos, etc.)

---

## 4. Enforce `go test -race` in CI

Currently, many data races go undetected because `-race` is not mandatory in CI. Given the scale of race conditions found (#3247), this is the single most impactful process change.

**Action items:**
1. Add `-race` flag to CI test runs (at minimum for core packages)
2. Fix all existing race detector failures
3. Block merges on race detector failures

---

## 5. API hygiene — stop exposing internal state

Several public APIs return internal references that callers can mutate, breaking encapsulation:

| API | Returns | Risk |
|-----|---------|------|
| `URL.GetParams()` | Internal `url.Values` map | Caller mutation + race |
| `GetDefaultConfigReader()` | Internal `defaults` map | Concurrent iteration crash |
| `GetRouterFactories()` | Internal `routers` map | Concurrent iteration crash |
| `GetProviderServiceMap()` | Internal `proServices` map | Concurrent iteration crash |
| `GetConsumerServiceMap()` | Internal `conServices` map | Concurrent iteration crash |
| `GetAllCustomShutdownCallbacks()` | Internal `*list.List` | Concurrent mutation |
| `GetCustomizers()` | Internal slice reference | Concurrent append+sort |

**Rule**: Public getters must return copies, never internal references. This should be a review checklist item.

---

## 6. Reduce `config` package coupling

The ongoing refactor (PRs #3231, #3232) to remove `config` package dependencies from protocol implementations is the right direction. But the approach needs more care:

- **Don't silently change error semantics** (e.g., `LoadRegistries` changing from panic to silent skip)
- **Don't break public APIs without deprecation** (e.g., `WithMethod` signature change)
- **Validate configuration at init time** (don't remove init-time validation without replacement)
- **Use URL attributes consistently** — document which attributes are expected and where they're set

---

## 7. Standardize error handling — eliminate `panic` in library code

The extension package uses `panic` for missing registrations:

```go
func GetProtocol(name string) base.Protocol {
    if protocols[name] == nil {
        panic("protocol for [" + name + "] is not existing, ...")
    }
}
```

This is inappropriate for a library/framework. A missing extension should return an error, not crash the entire process. The caller can decide whether to panic.

**Rule**: Reserve `panic` for truly unrecoverable programmer errors. Extension lookups should return `(T, error)`.

---

## Priority

| Priority | Item | Impact | Effort |
|----------|------|--------|--------|
| P0 | Fix `dubbo.go:301` wrong lock | Crash prevention | 1 line |
| P0 | Enable `-race` in CI | Prevent future regressions | Config change |
| P0 | `extension.Registry[T]` | Fix 20+ race-prone files | Medium (1 week) |
| P1 | URL `DeleteAttribute` + cleanup | Memory leak prevention | Small |
| P1 | Fix `RouterChain.Route` lockless read | Crash prevention | Small |
| P1 | Return copies from public getters | Crash prevention | Small |
| P2 | URL redesign | Long-term maintainability | Large |
| P2 | Lifecycle interface | Leak prevention | Medium |
| P2 | Eliminate panics | API quality | Medium |

Happy to help with implementation of any of these items. Related issues: #3246, #3247.

### PR DESCRIPTION
**Pull request by xxs588**
### Description
Fixes #3247

This PR follows the concurrency-safety direction from #3248 and #3249, and focuses on remaining race-prone shared states without introducing directory-level reorganization.

Changes included:
- `registry/directory`: guard `cacheInvokers`, `configurators`, and `SubscribedUrl` access with locks/snapshots
- `registry/protocol`: synchronize `overrideSubscribeListener.configurator` read/write
- `filter/hystrix`: snapshot regex cache under lock before callback execution
- `metrics`: protect application info globals via `sync.RWMutex`

Atomic commits:
1. `fix(directory): guard registry directory concurrent state access`
2. `fix(registry/protocol): synchronize override listener configurator`
3. `fix(filter/hystrix): avoid race when reading regex cache`
4. `fix(metrics): guard app info global state with RWMutex`

Key local verification:
- `make lint`
- `go test -race ./registry/directory -run 'TestRegistryDirectory(IsAvailableConcurrentWithCacheUpdates|ConfiguratorConcurrentAccess|SubscribedURLConcurrentAccess)$'`
- `go test -race ./registry/protocol -run 'TestOverrideSubscribeListenerConfiguratorConcurrentAccess$'`
- `go test -race ./filter/hystrix -run 'TestHystrixFilterConcurrentRegexAccess$'`
- `go test -race ./metrics -run 'TestAppInfoConcurrentAccess$'`

### Checklist
- [x] I confirm the target branch is `develop`
- [x] Code has passed local testing
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: filter/hystrix/filter.go [hunk 1]
```diff
@@ -162,6 +162,7 @@ func (f *Filter) Invoke(ctx context.Context, invoker base.Invoker, invocation ba
 		configLoadMutex.Unlock()
 	}
 	configLoadMutex.RLock()
+	methodRegexps := append([]*regexp.Regexp(nil), f.res[invocation.MethodName()]...)
```
Comment (AlexStocks) [inline filter/hystrix/filter.go]: 这里在 RUnlock 之前做了 map copy，然后在 Do 的回调里用 copy 没问题（避免了迭代时的 race）。但 `hystrix.Do` 的回调是异步执行的，hystrix 内部维护的 circuit 状态在 Do 返回后仍可能变化，和 filter 层的 goroutine 安全是两回事。

这个是原有设计问题，不是这次引入的。hystrix filter 本身整体是 non-thread-safe 的，注释里应该加一句说明这个 filter 不支持并发调用，或者考虑在 Invoke 入口加锁。

Comment (AlexStocks) [inline filter/hystrix/filter.go]: copy 语义正确，但建议明确注释说明这是在释放锁之前做 copy，避免后续维护者误以为可以在锁外再做 copy：
```go
// Must copy before unlock to avoid concurrent map read/write during iteration
methodRegexps := append([]*regexp.Regexp(nil), f.res[invocation.MethodName()]...)
```

**Review by AlexStocks — COMMENTED** [review]
继续修复 #3247 的 race，方向没问题，涉及多个包的全局状态保护。有几个问题需要关注，详见行内评论。

File: config_center/file/listener.go [hunk 2]
```diff
@@ -151,12 +187,9 @@ func (cl *CacheListener) Close() error {
 func (cl *CacheListener) AddListener(key string, listener config_center.ConfigurationListener) {
```
Comment (AlexStocks) [inline config_center/file/listener.go:189]: listenerSet.Snapshot() 在 RLock 里面做了 copy，但返回的 []ConfigurationListener 里每个元素是指针。如果外部对这些 listener 调用了非线程安全的操作，仍然会有问题。不过这是调用方的问题，不是 snapshot 本身的问题。

建议在 Snapshot() 的注释里说明返回的是只读快照，调用方不应修改 snapshot 里的 listener。

File: metrics/common.go [hunk 3]
```diff
@@ -43,19 +47,25 @@ type ApplicationMetricLevel struct {
 	HostName        string
 }
 
-var applicationName string
-var applicationVersion string
+var (
+	appInfoLock        sync.RWMutex
+	applicationName    string
+	applicationVersion string
+)
```
Comment (AlexStocks) [inline metrics/common.go:55]: InitAppInfo 写全局变量有锁，但 NewConfigCenterLevel 里调 getAppInfo() 是只读没问题。不过 InitAppInfo 只在应用启动时调用一次（正常路径），而 GetApplicationLevel 在每次上报 metric 时都可能调用，这里用 RLock 是对的。

有个潜在问题：如果 InitAppInfo 写入了不完整的状态（比如只写了 name 没写 version），GetApplicationLevel 读到的是新 name + 旧 version 的组合。不过这是逻辑问题，不是 race 问题。建议在 InitAppInfo 里把两个赋值放在一次锁里面，已经是这样了，没问题。

File: config_center/file/listener.go (hunk 2 above)
Comment (xxs588) [reply config_center/file/listener.go:189]: 图片
感谢，已加上

File: filter/hystrix/filter.go (hunk 1 above)
Comment (xxs588) [reply filter/hystrix/filter.go]: 图片
已加上，感谢

File: metrics/common.go (hunk 3 above)
Comment (xxs588) [reply metrics/common.go:55]: 图片
完善了注释

**Comment by xxs588** [comment]
修复了一下冲突，先 fetch了develop，然后对当前分支做了 rebase。  
过程中遇到的冲突是预期内的  modify/delete（filter/hystrix/），因为 develop分支已经把这部分迁移并删除了。  
然后我保持 develop分支 的删除结果，移除本分支对应的 hystrix 改动，然后继续变基，只保留本 PR 本身要交的逻辑

另外我排查到一个并发测试点  directory.go  里已经有  snapshotCacheInvokers()  这个并发安全读路径，但是测试里有直接读取 cacheInvokers 的写法 这样子会绕过同步语义并触发 race。  
但是这个属于测试侧读取模型问题，不应该在 pr 3269 里面修复所以为了保持 PR 原子性，我切到 split/3247-registry-lifecycle 单独提交一个最小测试补丁，只修复了一下测试读取路径

File: registry/directory/directory_test.go [hunk 4]
```diff
@@ -226,3 +231,112 @@ func TestToGroupInvokers(t *testing.T) {
 		assert.Len(t, registryDirectory.toGroupInvokers(), 2)
 	})
 }
+
+func TestRegistryDirectoryIsAvailableConcurrentWithCacheUpdates(t *testing.T) {
```
Comment (AlexStocks) [inline registry/directory/directory_test.go:342]: 这里补了并发场景用例，但当前包在 `go test -race ./registry/directory` 下还是会报 race。触发点不是这几个新 helper，而是这个测试文件里原有用例还在直接读 `registryDirectory.cacheInvokers`（例如 `TestSubscribe`、`Test_Destroy`、`Test_RefreshUrl`）。既然这个 PR 的目标是把 directory 的共享状态 race 收干净，建议把这些断言也改成通过 `snapshotCacheInvokers()` / `List()` 读取，或者在测试里统一走 `invokersLock`，否则目录包的 race 检测结果仍然是红的。

File: filter/hystrix/filter.go (hunk 1 above)
Comment (AlexStocks) [reply filter/hystrix/filter.go]: 这条问题目前还没处理。当前分支已经不再修改 `filter/hystrix/filter.go`，所以这次 race 修复没有覆盖到这里；如果这个 PR 不准备把 hystrix 一并收进来，建议至少后续单独补一个修复，把并发前提写清楚，或者在入口做串行化保护。

File: registry/directory/directory_test.go (hunk 4 above)
Comment (xxs588) [reply registry/directory/directory_test.go:342]: 感谢提醒，我现在已经把把 TestSubscribe、Test_Destroy、Test_MergeProviderUrl、Test_MergeOverrideUrl、Test_RefreshUrl 里直接读 cacheInvokers 的断言，全部改成了snapshotCacheInvokers() 读取，修复在 #3270  分支里面，go test -race ./registry/directory ./registry/nacos 也是通过的

Comment (xxs588) [reply registry/directory/directory_test.go:342]: 具体详情是
图片
图片
图片
图片
图片

File: filter/hystrix/filter.go (hunk 1 above)
Comment (xxs588) [reply filter/hystrix/filter.go]: filter/hystrix/filter.go 已在 #3253 迁移并从当前仓库删除，为保持原子性，我会在 hystrix 迁移后的目标位置单独跟进并发约束

**Comment by xxs588** [comment]
跟进同步：针对 https://github.com/apache/dubbo-go/pull/3269#discussion_r2985395111 提到的 hystrix 并发约束问题，已在迁移后目标仓建立后续 issue：https://github.com/apache/dubbo-go-extensions/issues/14 。
由于 hystrix filter 已在 #3253 从 dubbo-go 主仓迁移并删除，#3269/#3270/#3271 不再包含该文件改动。为保持 PR 原子性，这里不扩 scope 修改 hystrix 代码，后续将在扩展仓按 issue #14 单独推进（先明确并发约束，再评估是否需要入口串行化保护）。

File: filter/hystrix/filter.go (hunk 1 above)
Comment (xxs588) [reply filter/hystrix/filter.go]: 补充闭环：已在迁移后仓库创建跟进 issue：apache/dubbo-go-extensions#14。该问题后续在扩展仓推进，这个 PR 里不扩 scope。

**Review by AlexStocks — COMMENTED** [review]
## Code Review: fix(race): close remaining #3247 shared-state races

### [P0] 阻断级问题

**1. `registry/directory/directory.go:586-620` - 并发修复不完整**

新增的 `snapshotCacheInvokers()` 方法使用 RLock+copy 模式正确，但在 `toGroupInvokers()` 方法中仍直接访问 `dir.cacheInvokers`：

```go
func (dir *RegistryDirectory) toGroupInvokers() []protocolbase.Invoker {
    groupInvokersMap := make(map[string][]protocolbase.Invoker)
    // 问题：这里直接访问 dir.cacheInvokers，未使用 snapshot 方法
    for _, ivk := range dir.cacheInvokers {
        // ...
    }
}
```

需要统一修改为：
```go
for _, ivk := range dir.snapshotCacheInvokers() {
    // ...
}
```

### [P1] 规范级问题

**1. `config_center/file/listener.go:45-80` - listenerSet 设计**

新增的 `listenerSet` 结构设计良好，使用 `Snapshot()` 返回切片避免迭代时修改。但文档说明不足：

```go
// Snapshot returns a read-only listener snapshot for safe iteration.
// Callers must not mutate listeners through this snapshot.
```

建议添加注释说明调用方不应修改返回的切片。

**2. 测试辅助函数验证**

`listener_test.go` 新增的 `drainEvents` 和 `assertNoEvent` 函数未单独测试，建议添加单元测试验证其正确性。

---

总体评价：竞态修复方向正确，但需要检查所有直接访问共享状态的地方是否已统一使用 snapshot 方法。

**Comment by xxs588** [comment]
> ## Code Review: fix(race): close remaining #3247 shared-state races
> ### [P0] 阻断级问题
> 
> **1. `registry/directory/directory.go:586-620` - 并发修复不完整**
> 
> 新增的 `snapshotCacheInvokers()` 方法使用 RLock+copy 模式正确，但在 `toGroupInvokers()` 方法中仍直接访问 `dir.cacheInvokers`：
> 
> ```go
> func (dir *RegistryDirectory) toGroupInvokers() []protocolbase.Invoker {
>     groupInvokersMap := make(map[string][]protocolbase.Invoker)
>     // 问题：这里直接访问 dir.cacheInvokers，未使用 snapshot 方法
>     for _, ivk := range dir.cacheInvokers {
>         // ...
>     }
> }
> ```
> 
> 需要统一修改为：
> 
> ```go
> for _, ivk := range dir.snapshotCacheInvokers() {
>     // ...
> }
> ```
> 
> ### [P1] 规范级问题
> 
> **1. `config_center/file/listener.go:45-80` - listenerSet 设计**
> 
> 新增的 `listenerSet` 结构设计良好，使用 `Snapshot()` 返回切片避免迭代时修改。但文档说明不足：
> 
> ```go
> // Snapshot returns a read-only listener snapshot for safe iteration.
> // Callers must not mutate listeners through this snapshot.
> ```
> 
> 建议添加注释说明调用方不应修改返回的切片。
> 
> **2. 测试辅助函数验证**
> 
> `listener_test.go` 新增的 `drainEvents` 和 `assertNoEvent` 函数未单独测试，建议添加单元测试验证其正确性。
> 
> 总体评价：竞态修复方向正确，但需要检查所有直接访问共享状态的地方是否已统一使用 snapshot 方法。

关于 toGroupInvokers


我这边重新对了当前分支代码，这里现在是走 cacheInvokersMap.Range(...) 做分组，不是直接遍历 dir.cacheInvokers。cacheInvokers 目前只在加锁的 helper（snapshotCacheInvokers / swapCacheInvokers）里访问   然后剩下两个部分已补充注释和添加了相关单测

File: registry/directory/directory.go [hunk 5]
```diff
@@ -70,9 +70,11 @@ type RegistryDirectory struct {
 	consumerURL                    *common.URL
 	cacheOriginUrl                 *common.URL
 	configurators                  []config_center.Configurator
+	configuratorsLock              sync.RWMutex
 	consumerConfigurationListener  *consumerConfigurationListener
 	referenceConfigurationListener *referenceConfigurationListener
 	registerLock                   sync.Mutex // this lock if for register
+	subscribedURLLock              sync.RWMutex
```
Comment (Alanxtl) [inline registry/directory/directory.go]:
```suggestion
	subscribedUrlLock              sync.RWMutex
```

**Review by nanjiek — APPROVED** [review]
lgtm

**Comment by Alanxtl** [comment]
pls fix confilct

**Review by Alanxtl — APPROVED** [review]
lgtm
